.github/workflows/unittests-gpu.yml

```diff
@@ -39,17 +39,62 @@ jobs:
+                                             --work-dir tools/batch \
+                                             --save-path temp \
+                                             --remote https://github.com/${{ github.repository }} \
+                                             --command "./batch_states/test.sh | tee > ./script.log" \
+                                             --wait | tee > temp.log  
+
 
       - name: Test project on AWS Batch(For pull request)
         if: ${{ github.event_name == 'pull_request' || github.event_name == 'pull_request_target' }}
         run: |
-          python ./tools/batch/submit-job.py --region us-east-1 --job-type g4dn.4x --source-ref ${{ github.event.pull_request.head.ref }} --work-dir tools/batch --remote https://github.com/${{ github.event.pull_request.head.repo.full_name }} --command "./batch_states/test.sh" --wait | tee > script.log
+          echo "Start submitting job"
+          python ./tools/batch/submit-job.py --region us-east-1 \
+                                             --job-type g4dn.4x \
+                                             --name GluonNLP-${{ github.ref }}-${{ github.run_id }} \
+                                             --source-ref ${{ github.event.pull_request.head.ref }} \
+                                             --work-dir tools/batch \
+                                             --save-path temp \
+                                             --remote https://github.com/${{ github.event.pull_request.head.repo.full_name }} \
+                                             --command "./batch_states/test.sh | tee > ./script.log" \
+                                             --wait | tee > temp.log
+
+      - name: Wait for job and copy files from AWS s3
+        if: ${{ failure() || success() }}
+        run: |
+          head -100 temp.log | grep -oP -m 1 'jobId: \K(.*)' > jobid.log
+
+          echo "Job ID is"
+          cat jobid.log
+
+          cat jobid.log | xargs -i python ./tools/batch/wait-job.py --job-id {}
+
+          echo "Copy Codecov file"
+          cat jobid.log | xargs -i aws s3api wait object-exists --bucket gluon-nlp-dev --key batch/{}/temp/coverage.xml
+          cat jobid.log | xargs -i aws s3 cp s3://gluon-nlp-dev/batch/{}/temp/coverage.xml ./coverage.xml
+
+          echo "Copy log file"
+          cat jobid.log | xargs -i aws s3api wait object-exists --bucket gluon-nlp-dev --key batch/{}/temp/script.log
+          cat jobid.log | xargs -i aws s3 cp s3://gluon-nlp-dev/batch/{}/temp/script.log ./script.log
+
+      - name: Upload coverage to Codecov
+        if: ${{ failure() || success() }}
+        uses: codecov/codecov-action@v1.0.10
+        with:
+          env_vars: OS,PYTHON
+          file: ./coverage.xml
 
       - name: Upload log file for AWS Batch test results
+        if: ${{ failure() || success() }}
         uses: actions/upload-artifact@v2
         with:
```

Comment:
Maybe have an output folder where tests can dump outputs to, and then always expose that folder as artifact? This should help simplify maintenance for managing build artifacts.

Reply:
You mean redirect the test results as stdout and stderr to different files in a folder and then upload  this folder as build artifact so that the developers can access it through github workflow's artifacts.

Reply:
yes exactly.

Reply:
For me, I think both look fine and we may merge this is first and improve later.

Reply:
> You mean redirect the test results as stdout and stderr to different files in a folder and then upload this folder as build artifact so that the developers can access it through github workflow's artifacts.

> yes exactly.

That's the existing practice in 0.x but doesn't work well as it doesn't capture errors outside the immediate test execution. Instead, we need to capture the full cloudwatch log.